Refactor docking

royfalk opened the pull request:
- Move some functionality to dock_utils.
- Rename CanDockWithMe to CanDock and modify it.
- DistanceTwoTargets now takes both radii into account every time.
- Docking to a planet now uses its own configuration and does not rely on autopilot.
- Introduce simple dock to non-planets. Can dock within 5000m.
- Refactor docking text.
- Refactor _PretyDistanceString_ function.
- Rename _abletodock_ to _PlayDockingSound_ for the obvious reason it describe the function better.


Code Changes:
- [ ] Have the PR Validation Tests been run? See https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation. _Extensive testing of docking only._ 


Issues:
- VDU has two different texts. The top one in grey is static and is one of four(?) options. The bottom is yellow and dynamic. They provide pretty much the same info. Need to refactor but this is significantly harder.

kheckwrecker commented:
Tested this with a fresh install, starting a new campaign. Launched my Llama and flew from Atlantis to Serenity and the new docking functionality worked perfectly. The distance to dock started displaying well out from the station and stayed displayed until the Dock:Ready message appeared at 5000m as noted above. As soon as it said Dock:Ready, I was able to dock without issue. However, the return flight to Atlantis did not go as expected. The distance to dock started displaying well out from the plante, and the Dock:Ready message appeared at 5000m, but I was not able to actually dock until 1000 m out (which is what the distance has always been for Atlantis). 

Also of note, the distances between planets are different now with this PR merged. Atlantis to Serenity are now 5.90 light seconds apart, where it used to be something like 98 light minutes. The distance counts down slowly like it used to when it was light minutes, even though it says light seconds. When travelling at 97.4C, the light seconds used to count down very quickly. Also, when it switches from light seconds to kilometers, the distance in kilometers is much larger than it ever was. I'm not sure any of this is an problem or not - maybe the old distance numbers were wrong? I thought it best to report it all the same, since it is different than what it used to be.

Finally, not new to this PR, but I've never heard a docking sound play. I didn't realize there even was one. This is probably related to issue #1015, which I reported previously.

yowshi commented:
every last piece of equipment seems to have 
error: pyobject_callobject is null

royfalk commented:
**Atlantis range to dock**
Radius = 5000 km
dock_planet_radius_percent = 1.5
Range to dock: radius * percent = 7500
I just checked and it's right on the money.

> The distance to dock started displaying well out from the planet

It starts at 100km from docking point for both planets and other.

> Also of note, the distances between planets are different now with this PR merged. Atlantis to Serenity are now 5.90 light seconds apart,

This is a bug I introduced. I divided by 1000 to get km. 5.9 * 1000 = 98 light minutes.

> every last piece of equipment seems to have
Error: PyObject_CallObject is null

You are probably on the 0.9 assets. If not, please report the stdout/console/log of the error. It'll tell me if it's missing assets or a bug of some sort.

royfalk commented:
> every last piece of equipment seems to have
    Error: PyObject_CallObject is null

~You are probably on the 0.9 assets. If not, please report the stdout/console/log of the error. It'll tell me if it's missing assets or a bug of some sort.~
Error is:
```

Error: PyObject_CallObject failed
Traceback (most recent call last):
  File "/home/roy/git/Assets-Production/python/base_computer/upgrade_view.py", line 179, in get_upgrade_info
    key = key.replace('__upgrades','')
          ^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'replace'
```

I'll try and fix this and #1038 today.

BenjamenMeyer reviewed (approved):
Love this clarity

yowshi commented:
> **Atlantis range to dock** Radius = 5000 km dock_planet_radius_percent = 1.5 Range to dock: radius * percent = 7500 I just checked and it's right on the money.
> 
> > The distance to dock started displaying well out from the planet
> 
> It starts at 100km from docking point for both planets and other.
> 
i note the actual dockable radius of atlantis to be 7500 KM so that holds


> > every last piece of equipment seems to have
> > Error: PyObject_CallObject is null
> 
> ~You are probably on the 0.9 assets. If not, please report the stdout/console/log of the error. It'll tell me if it's missing assets or a bug of some sort.~ Error is:
> 
> ```
> 
> Error: PyObject_CallObject failed
> Traceback (most recent call last):
>   File "/home/roy/git/Assets-Production/python/base_computer/upgrade_view.py", line 179, in get_upgrade_info
>     key = key.replace('__upgrades','')
>           ^^^^^^^^^^^
> AttributeError: 'dict' object has no attribute 'replace'
> ```
> 
> I'll try and fix this and #1038 today.

i checked with a freshly downloaded and unzipped assets-production-master, i am not getting this error.

yowshi commented:
on a game play feel note. i love the new docking radia, even the station docking radius of 5KM. though i also kind of liked the station docking minigame.

kheckwrecker commented:
The distance is now correct and performs as expected when it's showing light minutes and light seconds. However, when I'm close enough to a target to have it show in kilometers, the number displayed is incorrect. For example, when I launch from Atlantis, the distance to Atlantis (right after launching) is 9,222,001Km (but if I turn my ship around for a visual, the planet is right there, close as ever) and the distance to Ataraxia is 700,067,133Km. It's almost like the decimal place is off by 1,000 when the distance is shown in kilometers. All the other planets and bases, which are light minutes or light seconds away show the correct distance. 

When I flew back to from Serenity to Atlantis to dock, the distance showed as approx. 7,500,000 km when I was able to dock, which further leads me to believe the decimal place is off by 1,000.

I'm now having trouble with autopilot (unable to smoothly approach planet - keep activating and deactivating and overshooting the planet) and the effect of gravity wells seems to have disappeared also, but I'm thinking these are both caused by the incorrect distance in kilometers?

kheckwrecker commented:
> **Atlantis range to dock** Radius = 5000 km dock_planet_radius_percent = 1.5 Range to dock: radius * percent = 7500 I just checked and it's right on the money.
> 
> > The distance to dock started displaying well out from the planet
> 
> It starts at 100km from docking point for both planets and other.
> 
Thank you for the clarification. So docking at Atlantis used to be at 1,000 km, but it's now at 7,500 km? I'm definitely not objecting to the change. In fact, I rather like it.

One thing to note (maybe for a future patch), but given the extended docking distances, the docking squares that show when you request docking are now irrelevant, since ships can dock well before reaching those squares. 

Again, in case I wasn't clear, I really like the way docking works now. Thank you!

yowshi commented:
> **Atlantis range to dock** Radius = 5000 km dock_planet_radius_percent = 1.5 Range to dock: radius * percent = 7500 I just checked and it's right on the money.
> 
> > The distance to dock started displaying well out from the planet

can confirm 7500 KM is atlantis's new docking radius. docking range for atlantis starts displaying at 7600 KM

yowshi commented:
there is definitly something up with serenity though. it's shields never recharge. may or may not have something to do with the fact my shields take damage on every trip i take.

kheckwrecker commented:
> the integral part duplication is slowly adding mass to my ship. attempting to sell my repair droid as part of a test to watch this duplication happen with debug=3 on caused a glitch and now i have 0.086 of 360 cubic meters for upgrade space and nothing i sell increases my equipment space. oops.
> this is happening even on a fresh new campaign and when i buy a plowshare and sell one of it's equipment.

This part should be reported under issue #1038 rather than here, as it's a separate issue.

yowshi commented:
> > the integral part duplication is slowly adding mass to my ship. attempting to sell my repair droid as part of a test to watch this duplication happen with debug=3 on caused a glitch and now i have 0.086 of 360 cubic meters for upgrade space and nothing i sell increases my equipment space. oops.
> > this is happening even on a fresh new campaign and when i buy a plowshare and sell one of it's equipment.
> 
> This part should be reported under issue #1038 rather than here, as it's a separate issue.

i have moved it.

vincele commented on engine/src/cmd/dock_utils.cpp:
```diff
@@ -0,0 +1,200 @@
+    } else {
+        if(configuration()->dock.simple_dock && target->pImage->dockingports.size() != 0 && 
+            range < 100000) {
+            if (range > 5000) {
```
These numbers should be made constants, so that they are defined only once in a single place, this will avoid discrepancies, if only one instance is changed in the future. That would also make them easier to make them configurable, if we want that.

royfalk replied:
I agree. But I was working under a time pressure to get a fix in 24 hours. I'll make the changes.

royfalk replied:
Fixed.

vincele commented on engine/src/cmd/dock_utils.cpp:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    // Distance in km
+    static const double light_second = 299792458;
```
Shouldn't all these constants be defined more globally ?
`engine/src/physics.h` looks appropriate.

royfalk replied:
Good catch. It's actually defined there.

kheckwrecker replied:
I'm confused. When I google the distance of a light second in kilometers, I'm told it's 299,792.458, not 299792458. Is this divided by 1,000 somewhere in the code?

royfalk replied:
Distance in the game is always in meters. We convert it to other units for display only.

royfalk replied:
Code refers to physics.h now.

vincele commented on engine/src/cmd/dock_utils.h:
```diff
@@ -0,0 +1,37 @@
+std::string GetDockingText(Unit *unit, Unit *target, double range);
+std::string PrettyDistanceString(double distance);
+
+#endif // VEGA_STRIKE_ENGINE_CMD_DOCK_UTILS_H
```
Missing CRLF @ EOF

vincele commented on engine/src/cmd/dock_utils.cpp:
```diff
@@ -0,0 +1,200 @@
+
+    // Use light years
+    return (boost::format("%.2lf light years") % (distance / light_year)).str(); 
+}
```
Missing CRLF @ EOF

vincele commented on engine/src/cmd/unit_generic.cpp:
```diff
@@ -201,22 +202,22 @@ void Unit::Ref() {
+extern void PlayDockingSound(int dock);
 
-bool CrashForceDock(Unit *thus, Unit *dockingUn, bool force) {
+/*bool CrashForceDock(Unit *thus, Unit *dockingUn, bool force) {
```
Why keep it commented ?

royfalk replied:
Welcome to Vegastrike. There should probably be a document somewhere (@BenjamenMeyer ), or maybe not.
I don't know how much you know about the game.
It was originally developed by Daniel Horn and a couple of other people in 1999-2005 or so. Somewhere along the way, I think efforts shifted toward making the Privateer assets work on the engine. 
And then people moved on and development sort of stopped.
In 2019 a couple of people came together and restarted development. But the technical debt due to time, loss of knowledge and plenty of other reasons was significant. 
`CrashForceDock` is something from the original devs. Like the comment _"We cannot dock if we are already docked"_. I try not to remove these unless I know for sure it shouldn't stay. As this PR is a quick bug fix, I'm keeping changes to a minimum.
However, as I noticed this function is unused, I commented it out.

vincele replied:
Eh, either you're keeping changes to a minimum or you do additionnal things, like commenting out dead code... ;-)

I already worked a bit on VS in the (distant) past, and keep comig back to it, but I know almost nothing about the code. All my comments are from this naive perspective. Don't take them as harsh or anything. They're mostly just questions.

I'm not saying that you're wrong, but reviews are easier if you keep the changes to the bare minimum. This IMHO should be done in another cleanup PR, or at least a separated commit that does just this "remove dead code". There's no need to keep it, it is in the git history anyways. As a separate PR this trivial thing can be merged easily and independently of the bug fix.

royfalk replied:
> Eh, either you're keeping changes to a minimum or you do additionnal things, like commenting out dead code... ;-)

I know. The temptation.

> I already worked a bit on VS in the (distant) past, and keep comig back to it, but I know almost nothing about the code. 

@BenjamenMeyer can give you dev privileges, if you'd like.


> I'm not saying that you're wrong, but reviews are easier if you keep the changes to the bare minimum. This IMHO should be done in another cleanup PR, or at least a separated commit that does just this "remove dead code". 

I agree. Unfortunately, due to the excessive technical debt, we have some truly monstrous PRs. And there is a cost to doing things cleanly. I'm already juggling several PRs, in various stages of readiness.

> There's no need to keep it, it is in the git history anyways. 

This is true for people with functioning memories. I introduced a bug by merging an engine PR and forgetting there's a complementing assets one.

stephengtuggy replied:
I have to agree with @vincele here. As a general rule, do not make functional changes in the same PR with refactor changes. It makes for huge, difficult-to-follow PRs, and a nightmare in case we need to go back and revert any of the functionality changes. You've done way too much of this in the past, and I am not happy about it. I should have reviewed your past PRs more carefully.

And if you're going to take out sections of code, don't bother commenting them out. Just remove them completely -- at an appropriate time, i.e. in their own PR.

vincele commented on engine/src/configuration/configuration.cpp:
```diff
@@ -369,9 +370,7 @@ void Configuration::OverrideDefaultsWithUserConfiguration() {
-    physics_config.planet_dock_port_size = GetGameConfig().GetFloat("physics.planet_port_size", physics_config.planet_dock_port_size);
-    physics_config.planet_dock_port_min_size = GetGameConfig().GetFloat("physics.planet_port_min_size", physics_config.planet_dock_port_min_size);
-    physics_config.jump_mesh_radius_scale = GetGameConfig().GetFloat("physics.jump_mesh_radius_scale", physics_config.jump_mesh_radius_scale);
+     physics_config.jump_mesh_radius_scale = GetGameConfig().GetFloat("physics.jump_mesh_radius_scale", physics_config.jump_mesh_radius_scale);
```
stray whitespace change ?

royfalk replied:
Done.

vincele commented on engine/src/configuration/configuration.h:
```diff
@@ -380,6 +380,7 @@ struct PhysicsConfig {
     float max_lost_target_live_time{30.0F};
     float percent_missile_match_target_velocity{1.0F};
     double game_speed{1.0};
+    bool game_speed_lying{false};
```
Why is this change here ?
looks unrelated to the PR

royfalk replied:
This is used by `PrettyDistanceString`, which I refactored. I also moved it from the xml configuration (vegastrike.config) to the new config.json.

vincele replied:
OK, I missed that

vincele commented on engine/src/gfx/cockpit.cpp:
```diff
@@ -679,8 +680,12 @@ float GameCockpit::LookupUnitStat(int stat, Unit *target) {
+                }
+
+                if (CanDock(station, target, true) != -1) {
+                    if (CanDock(station, target, false) != -1) {
```
This reads a bit strange, maybe a comment could help a casual reader

royfalk replied:
Huh. The original code was:
             ```
   if (station->CanDockWithMe(target, true) != -1) {
                    if (station->CanDockWithMe(target, false) != -1) {
```
I've no idea why it this so.

vincele commented on engine/src/gfx/vdu.cpp:
```diff
@@ -933,10 +883,7 @@ void VDU::DrawNav(GameCockpit *cp, Unit *you, Unit *targ, const Vector &nav) {
-                            targ) : 0.0))
-                            .
-                                    str);
+                    + PrettyDistanceString(((you && targ) ? DistanceTwoTargets(you,targ) : 0.0));
```
missing space after the comma

vincele reviewed (commented):
Not sure if I'm fit for approving, so take those comments as just that, random comments from a casual reviewer

royfalk commented:
> Not sure if I'm fit for approving, so take those comments as just that, random comments from a casual reviewer

You are most welcome to provide comments. I generally strive to get a few yays, especially from the testers as they find all the nasty surprises left in the code.

royfalk commented on engine/src/cmd/dock_utils.cpp:
```diff
@@ -0,0 +1,200 @@
+
+    if (ignore_occupancy) {
+        for (unsigned int i = 0; i < dock->pImage->dockingports.size(); ++i) {
+            if (!dock->pImage->dockingports[i].IsOccupied()) {
```
You can see the original code in unit_generic.cpp::2101. I don't fully understand this code but I think this simply checks **you** are in one of the docks. The following mode complex code checks if another ship is there and prevents you from docking. I simply reversed the order of the two.

vincele replied:
Ok, this is also an example of the "bare minimum to fix", see my other answer below, about commenting dead code.

You moved the code and at the same time changed it, so I missed that fact in the review, and my comments are off.

royfalk commented on engine/src/cmd/dock_utils.cpp:
```diff
@@ -0,0 +1,200 @@
+                                Transform(dock->GetTransformation(),
+                                        dock->pImage->dockingports[j].GetPosition().Cast())),
+                        dock->pImage->dockingports[j].GetRadius())) {
+                    // We cannot dock if we are already docked
```
Again, this is legacy code and comment.

vincele replied:
OK, then please disregard my question

royfalk commented on engine/src/cmd/dock_utils.cpp:
```diff
@@ -0,0 +1,200 @@
+
+        range -= target->rSize() * configuration()->dock.dock_planet_radius_percent;
+        if (range < 0) {
+            // Should be caught by dock check above
```
Exactly. I sometimes leave bombs like this in the code because I don't  fully understand it and I am sometimes surprised. Honestly, it's the same code here and in CanDock, so abort should never be called. 
If I'm wrong, one of the testers would likely have reported something.
I'll remove it before I merge.

royfalk replied:
Removed. It did crash the game for jump points.

vincele replied:
I was just arguing to change the comment to say:
`// Should be caught by the CanDock() check above`

kheckwrecker commented:
Did further testing tonight on this PR, with the following results:

Running a version of master with PR#1036, 1039, 1047 and 1048 cherry-picked and merged. Started a new campaign and launched from Atlantis. Distance from Atlantis showed as 9221980 kms (which doesn't look right). Distances to all other nav points looked correct (some in light minutes and some in light seconds) except for Ataraxia which showed as 70067125 kms.

Flew from there to Serenity. The distance to docking countdown started at 100000 kms (which seems high) and the docking ready indicator showed up at 5000 (which is perfect). Interestingly enough, as it was a patrol mission, I had to get to 200 meters from Phillies before the scanning would complete (which is a lot closer than it used to be).

Flew from Serenity to Phillies. Distance to target dropped as expected from light minutes thru light seconds but when it dropped to kilometers the distance was 299792458 kms. I got to about 260000 kms and the distance to docking countdown came on showing 100000 meters, but Phillies blew up in front of me so I couldn't finish that. However, I assume the docking ready would have shown at 5000 meters (as it did for approaching Serenity)

Flew from there to Plainfield (a relay station). Again, dropped out of light seconds 299792458 kms away (which is too high). I got to 100000kms way when the distance to docking count down started (seems too high). Docking ready indicator came on at 5000 meters (which is perfect). Unfortunately, I was approaching closer in order to scan the station (for my patrol mission) when the game froze for a second or two and then crashed to the desktop. 

I'll try again tomorrow. I especially want to test approaching Atlantis as I wasn't sure about the distances the last time I played and I want to make sure I have an accurate record.

yowshi commented:
> Flew from there to Plainfield (a relay station). Again, dropped out of light seconds 299792458 kms away (which is too high). I got to 100000kms way when the distance to docking count down started (seems too high). Docking ready indicator came on at 5000 meters (which is perfect). Unfortunately, I was approaching closer in order to scan the station (for my patrol mission) when the game froze for a second or two and then crashed to the desktop.
> 
> I'll try again tomorrow. I especially want to test approaching Atlantis as I wasn't sure about the distances the last time I played and I want to make sure I have an accurate record.

definitly a multiplication issue on 299792458. that is the exact speed of c in game. presumale in meters per second.

however it also appears here.

>         } else if (distance
>                 < 299792458) {                          //use kilometers without decimals up to 299792.458 km
>             sprintf(qr.str, "%.0lf kilometers", distance / 1000);

line 580 in Vega-Strike-Engine-Source/engine/src/gfx/vdu.cpp of at least the master branch of the engine source.
dropping out of light seconds is supposed to happen at this distance. assuming you didnt also drop out of spec?

royfalk commented:
> Did further testing tonight on this PR, with the following results:

> Distance from Atlantis showed as 9221980 kms 

Fixed issue where km did not divide by 1000.

> Flew from there to Serenity. The distance to docking countdown started at 100000 kms 

This is the new number. It is configurable.

> Interestingly enough, as it was a patrol mission, I had to get to 200 meters from Phillies before the scanning would complete (which is a lot closer than it used to be).

It actually wasn't. I changed the way distance is calculated. In the old way, distance was between centers except for planets. These subtracted their radius. I reverted this as it looks weird, especially when flying large ships.
Note that the 400 clicks in the patrol mission text is wrong, but unrelated to this PR.

> Phillies blew up in front of me so I couldn't finish that. 

One reason to disable planet damage. 

> Unfortunately, I was approaching closer in order to scan the station (for my patrol mission) when the game froze for a second or two and then crashed to the desktop.

I think I fixed that crash.

kheckwrecker reviewed (commented):
From a play testing perspective, this PR fixes almost all of the problems/challenges with docking. I was not able to dock with any jump points, the distances between planets are back to correct, and distance to docking and the appearance of Docking:Ready are correct, with one exception. The distance to docking countdown for Atlantis starts well out, and the Docking: Ready indicator appears at around 2300 km. However, you can't actually dock until 1,000 km out. This is the distance that it's always been, but I thought it was supposed to be changed to 7500 km out for Atlantis? 

Another interesting impact of this PR is that ships now launch further away from planets, which means that gravity wells no longer affect ships accelerating away from planets. I'm not complaining, but thought it worth noting.

kheckwrecker commented:
Update to my last comment, upon further testing....

Docking at stations (like Serenity, Plainfield and Ataraxia) is perfect. 5000 km every time, with the countdown to docking starting 100 km out at 5100 km, and a Docking: Ready message at 5000 km. 

Docking at planets is not quite perfect yet. Flying into Atlantis, the countdown to docking started at 7,600 km with a 100 km countdown for a theoretical docking at 7500 km. At this point everything was as expected. Unfortunately, the countdown ended at 7500 km, but no Docking: Ready message appeared. Finally was able to dock at 3,697 km, but there was no Docking: Ready message at any point. I just hit the "d" key every km on the way in until it docked.

As an Interesting side note, I also discovered that if you launch right away after landing, your launch distance is identical to the distance you docked at. So I launched at 3697 km, turned right around to land and tried to land, but could not. So I flew closer and the docking ready countdown came on at 1107 km and I was able to land at 1007 km from the planet. This time it did say Docking: Ready, so I landed as soon as I saw it. Then I launched right away again (at 1007 km from the planet), turned around and flew to 998 km from the planet (it did say Docking: Ready all the time) and landed. Launched again at 998 km from the planet this time. I'm not sure the launch distance thing is relevant, so I only mention it in passing. The real issue for me is the inconsistencies in the docking countdown and lack of  Docking: Ready message when approaching planets.

Next I flew to the planet Wiley. The docking countdown started at 5447 km out and lasted for 100 km. It ended at 5347 km, with no Docking: Ready message, and I could not dock. I was finally able to dock at 2024 km from the planet surface, but there was no Docking: Ready message at any point.

evertvorster commented:
Interesting. When you finally do get your docking ready message, don't dock, and wait a while. If the message goes away, then the planet's dock is busy, and that might explain why you were not able to dock straight away at the planets. 

There is an open bug on planets dock being busy, I forget where.

royfalk commented:
Let me save everyone the effort of reverse engineering my code.

**Planetary landing logic**
Distance = distance - planetary radius.
Docking ready = distance < 1/2 radius
Landing countdown starts at docking + radius and ends when ready.

**Atlantis**
Radius is 5m..
Displayed distance is actually minus 5m.
Countdown starts at 7.5m. Ends at 2.5m with dock ready as expected.

**Dock Busy** @evertvorster 
For planets, this is never true. 
For stations, only if _simple_dock_ isn't enabled. It is by default for now.
I promise the next thing is to expose the entire config for everyone to hack to their hearts content.

vincele commented:
For Atlantis, it's in meters ? That's *very* small. Is that not inducing problems ? (just naively wondering...)

royfalk commented:
> For Atlantis, it's in meters ? That's very small. Is that not inducing problems ? (just naively wondering...)

5 **million** meters. I'm looking at the code in the debugger. It's always in meters.

vincele commented:
OK, that's better... :-)

kheckwrecker commented:
This branch will no longer compile in Windows. So I cloned master and tried cherry picking the file changes in this PR, but there's a conflict in vdu.cpp between this branch and master. I tried picking the version of vdu.cpp that's in master, but compile generated this error message:
D:\VS_Test_Docking\Vega-Strike-Engine-Source\engine\src\gfx\vdu.cpp(718,69): error C2039: 'CanDockWithMe': is not a member of 'Unit' [D:\VS_Test_Docking\Vega-Strike-Engine-Source\build\VS2022Win64-pie-disabled-release\engine\vegastrike-
engine.vcxproj]
      D:\VS_Test_Docking\Vega-Strike-Engine-Source\engine\src\cmd\unit_generic.h(158,7):
      see declaration of 'Unit'

It also failed when I tried picking the version of vdu.cpp that is in this branch, but I was not quick enough to catch the error message.

stephengtuggy reviewed (dismissed):
I do not approve of this PR. You're oversimplifying things way too much, in my opinion. The game is supposed to be challenging and (at least somewhat) realistic.

kheckwrecker commented:
> I hate this PR. You're oversimplifying everything. The game is supposed to be challenging.
> 
> This isn't Vega Strike anymore. What you're making is a completely different game. That being the case, you should fork the repository, and make changes there, instead of screwing up the game engine's official repo.

Interesting observations. The PR started out due to the need to fix the broken distance to docking notification, so I think that's a good thing, as it fixes a really annoying bug. And, personally, I think a more consistent distance to docking, especially at stations is a good thing. I see two areas where the PR could be removing challenge. 1) If docking is further out, then launching is also probably further out, which reduces or eliminates the impact of gravity wells. I think gravity wells are boring, but evertvorster likes them. My original idea was a configuration setting that lets players choose. But, having discovered the shelton slide key, gravity wells are not such a big deal for me anymore, so I'm easy either way now. 2) Docking no longer requires players to fly their ships into little green triangles anymore in order to dock. Which is a reduction in the challenge of the game for sure, but I was okay with it as a trade-off for improved consistency in docking distances.

Are you seeing other changes that oversimplify?

stephengtuggy commented:
> > I hate this PR. You're oversimplifying everything. The game is supposed to be challenging.
> > This isn't Vega Strike anymore. What you're making is a completely different game. That being the case, you should fork the repository, and make changes there, instead of screwing up the game engine's official repo.
> 
> Interesting observations. The PR started out due to the need to fix the broken distance to docking notification, so I think that's a good thing, as it fixes a really annoying bug. And, personally, I think a more consistent distance to docking, especially at stations is a good thing. I see two areas where the PR could be removing challenge. 1) If docking is further out, then launching is also probably further out, which reduces or eliminates the impact of gravity wells. I think gravity wells are boring, but evertvorster likes them. My original idea was a configuration setting that lets players choose. But, having discovered the shelton slide key, gravity wells are not such a big deal for me anymore, so I'm easy either way now. 2) Docking no longer requires players to fly their ships into little green triangles anymore in order to dock. Which is a reduction in the challenge of the game for sure, but I was okay with it as a trade-off for improved consistency in docking distances.
> 
> Are you seeing other changes that oversimplify?

No, those are the changes that I am objecting to. And I apologize for my earlier comment. I was out of line. Not having such a good day for unrelated reasons.

kheckwrecker commented:
Well, yowshi and I both kind of miss the flying into the little green squares to dock. So, I'd be okay if that returned. You're still never going to convince me to like gravity wells however! And I'm sorry about your day - we've all had those, unfortunately.

yowshi commented:
> I do not approve of this PR. You're oversimplifying things way too much, in my opinion. The game is supposed to be challenging and (at least somewhat) realistic.

> > The PR started out due to the need to fix the broken distance to docking notification
> 
> Maybe the PR should be just that fix, and nothing more. It will be easier that way to only test, review & merge it without being distracted by the other changes.
> 
> Then the refactoring could be handled and discussed in another PR.
> 
> @royfalk what do you think, is it possible to extract the bare minimum fix in a separate simple single-commit PR ?

yeah same, you're never going to convince me that digging your way out of a planetary gravity well is part of the fun. it's simply a grind to me the sort of thing added to a game so a micro transaction could make it disappear.

kheckwrecker commented:
Interesting. I've been looking through the old open issues to see if any can be closed, and noticed that #37 has an extensive discussion about docking and possible changes. That issue was opened in 2020, so docking certainly isn't a new issue!

stephengtuggy commented:
@kheckwrecker You are correct. This subject has been under discussion for quite a while.

OK, how about if we change the docking behavior for planets only, and leave the station docking as-is for now? I would be OK with that. How about you guys?

kheckwrecker commented:
> @kheckwrecker You are correct. This subject has been under discussion for quite a while.
> 
> OK, how about if we change the docking behavior for planets only, and leave the station docking as-is for now? I would be OK with that. How about you guys?

Just to be clear what we're agreeing on::
1) Fix the bug in the distance to dock indicator?
2) Implementing a more logical formula for docking at planets, based on the math in this PR?
3) Leaving the docking model for stations as it currently is in the existing game (which includes needing to fly into the little green squares)?

If so, then I'm good with your suggestion.

Incidentally, there are two more old open issues on docking: #212 - Cannot Dock and #267 -  Docking indicator cuts in and out during approach

Probably should link those to PR so they close when the PR (in its final form) is merged with master?

yowshi commented:
> @kheckwrecker You are correct. This subject has been under discussion for quite a while.
> 
> OK, how about if we change the docking behavior for planets only, and leave the station docking as-is for now? I would be OK with that. How about you guys?

this sounds like awesome to me.

evertvorster commented:
Gravity wells are supposed to suck. Quite literally!
When you start with slow ships, there are supposed to be things that annoy you. Why else would you spend all that time and energy saving up for faster ships if there is no functional difference between the faster ship and the one you started out with?

It adds realism to the game. It is what sets Vega Strike apart from Oolite and similar games. 

As you learn a couple of tricks, even in the slower ships the gravity wells impact can be greatly mitigated, which is yet another reason to leave them as they are. 

Now what really does such without any hope of getting it better in the game is the fact that planetary docks can be occupied. It would be nice if one could purchase a priority pass in the game that allows you to dock when in range, and kicks out whoever is docked at the moment. :)

In my opinion, the docking distance should be a configurable option.

stephengtuggy commented:
I'm not sure if we know how to fix #212 and #267 yet. Unfortunately.

> In my opinion, the docking distance should be a configurable option.

I could go for that. Seems like everybody wins that way. We can probably introduce a setting specifically for the planet docking distance, and have it default to 7500 km like this PR already has it. Then, for now at least, we still require flying through the little green squares to dock with stations. To be revisited later, potentially.

> 1. Fix the bug in the distance to dock indicator?

If we know how to, sure.

> 2. Implementing a more logical formula for docking at planets, based on the math in this PR?

Yep.

> 3. Leaving the docking model for stations as it currently is in the existing game (which includes needing to fly into the little green squares)?

Yes, exactly.

yowshi commented:
> Gravity wells are supposed to suck. Quite literally! When you start with slow ships, there are supposed to be things that annoy you. Why else would you spend all that time and energy saving up for faster ships if there is no functional difference between the faster ship and the one you started out with?
> 
> It adds realism to the game. It is what sets Vega Strike apart from Oolite and similar games.
> 
> As you learn a couple of tricks, even in the slower ships the gravity wells impact can be greatly mitigated, which is yet another reason to leave them as they are.
> 
> Now what really does suck without any hope of getting it better in the game is the fact that planetary docks can be occupied. It would be nice if one could purchase a priority pass in the game that allows you to dock when in range, and kicks out whoever is docked at the moment. :)
> 
> In my opinion, the docking distance should be a configurable option.

try afterburning your way out of the atlantis gravity well in a mule, or into atlantis for a trade run for that matter. the mule is a bigger more expensive ship and if you fly one you will never want to land on a planet ever again. 

it is worth noting that wing commander privateer a spiritual predecessor to this game skipped docking sequences entirely and also skipped gravity well issues all together. and that wasnt a forced issue it was a design choice.

people saying that getting rid of the gravity well issue for docking will kill the game should maybe look back at some of the predecessors and their design choices. because some of them thought otherwise and they werent wrong.

royfalk commented:
# Point by Point

**@kheckwrecker** 
> This branch will no longer compile in Windows.

I'll take care of that. You really don't need to merge this or other PRs on top of master to test. Just test the branch itself.

**@stephengtuggy** 
> I do not approve of this PR. You're oversimplifying things way too much, in my opinion. The game is supposed to be challenging and (at least somewhat) realistic.

- I've made the simple_dock and option. Disable it and you get the same station docking as before. 
- Planet docking range did not change and was always configurable.
- I did tweak with the range code to make it work but it was buggy before. 

> I'm not sure if we know how to fix #212 and #267 yet. Unfortunately.

I'm pretty sure this fixes #267. I wasn't aware of 212 and did not touch the key mapping. However, I believe @evertvorster mentioned a how-to in the issue.

> And I apologize for my earlier comment. I was out of line. Not having such a good day for unrelated reasons.

All good. We've all had these.

**@evertvorster** 
> In my opinion, the docking distance should be a configurable option.

It is. 

> Really, the realism is what sets Vega Strike apart from the other "space
simulators".

I take no sides in this discussion of realism/fun, except where it impacts the code itself. Where possible, I make this configurable in the settings. 

> It is also worth noting that Vega Strike is a _different game_, with different goals. One of the main goals is realism.

Please separate the engine from the game. VS is realistic. The engine is supposed to support fun games like privateer as well.

royfalk commented:
# Proposal

- Make simple_dock disabled. People who want to dock to stations without flying into a dock should enable it pre-compilation for now, and in config.json when #1052 is merged.
- Actual distance to dock to planets is configurable. It is a factor of the planet radius. Atlantis radius is 5 million meters. That's why you're seeing these numbers. I picked 1.5, which is why you can dock at 2500.
- Distance to countdown is also configurable with the same value.  Countdown starts at 7500.
- Station docking - same as before (simple dock disabled) or a configurable 5k meters (with it).
- Station distance - 100km

What do you guys think?

Issues:
- There's a 5k hard coded that should change.
- The distance for stations only works for simple dock. Should change as well.

kheckwrecker commented:
> # Point by Point
> **@kheckwrecker**
> 
> > This branch will no longer compile in Windows.
> 
> I'll take care of that. You really don't need to merge this or other PRs on top of master to test. Just test the branch itself.
> 

That was the original problem. This branch will not compile due to the conflicts, so I have no way to test the branch itself. Merging PR's onto master was my attempt to get around this problem. Unfortunately, that didn't work either. I'll keep an eye out for your fixes to the conflicts.

yowshi commented:
> Gravity wells are supposed to suck. Quite literally! When you start with slow ships, there are supposed to be things that annoy you. Why else would you spend all that time and energy saving up for faster ships if there is no functional difference between the faster ship and the one you started out with?
> 
> It adds realism to the game. It is what sets Vega Strike apart from Oolite and similar games.
> 
> As you learn a couple of tricks, even in the slower ships the gravity wells impact can be greatly mitigated, which is yet another reason to leave them as they are.
> 
> Now what really does suck without any hope of getting it better in the game is the fact that planetary docks can be occupied. It would be nice if one could purchase a priority pass in the game that allows you to dock when in range, and kicks out whoever is docked at the moment. :)
> 
> In my opinion, the docking distance should be a configurable option.

try afterburning your way out of the atlantis gravity well in a mule, or into atlantis for a trade run for that matter. the mule is a bigger more expensive ship and if you fly one you will never want to land on a planet ever again. 

it is worth noting that wing commander privateer a spiritual predecessor to this game skipped docking sequences entirely and also skipped gravity well issues all together. and that wasnt a forced issue it was a design choice.

people saying that getting rid of the gravity well issue for docking will kill the game should maybe look back at some of the predecessors and their design choices. because some of them thought otherwise and they werent wrong.

> It feels like I have been over this a few times, but maybe I have not explained my position clearly enough.
> try afterburning your way out of the atlantis gravity well in a mule, or into atlantis for a trade run for that matter. the mule is a bigger more expensive ship and if you fly one you will never want to land on a planet ever again.
> Which is sort of the point. If you drive a tanker, you can expect it to be slow. Really, the realism is what sets Vega Strike apart from the other "space simulators". If the realism goes away, then so does the want to play Vega Strike. I would just play No Man's Sky or Elite Dangerous.. or Oolite. All of those have better graphics, and are way more polished, but none has the realistic feel of Vega Strike. Frankly, I have never understood the need to make all the games of a similar genre to play the same way. Variety is the spice of life, after all. To get back to your mule example, did you know that you can pay someone to do the tedious flying for you? You can just zip ahead in a small fast craft to find deals, then call the mule and trade. It costs a bit of money, but gets rid of the tedium of driving your tanker out of a deep gravity well. I have a youtube video on how to do this, if you are interested. In the beginning of the game, when you don't have a lot of cash, you have to deal with the inconveniencies of having a horribly overloaded Llama and deep gravity wells. It makes you re-evaluate your choices of cargo, etc.. it all adds to the realism of the game. Making this not suck makes the game boring. The quickest way out of a gravity well is to put it right at your back, hit "Y" to put your computer into travel mode, adjust thrust to fulle speed (You'll see it is three orders of magnitude faster than maneuver mode) hit Shift+A to switch on SPEC. And then wait a little. This is a good opportunity to look around, make sure that your destination is locked in missions are prepped that sort of thing. You can afterburn here, but it does not make much of a difference. Once the SPEC multiplier starts ticking it's really not a long time before you are out of the well and can switch to automatic navigation. Don't forget to switch your computer back to maneuver mode or you'll be likely to make a smear on the side of your destination...
> it is worth noting that wing commander privateer a spiritual predecessor to this game skipped docking sequences entirely and also skipped gravity well issues all together. and that wasnt a forced issue it was a design choice.
> It is also worth noting that Vega Strike is a _different game_, with different goals. One of the main goals is realism. Think of it more like "MS Flight Simulator for space", rather than Wind Commander, which are arguably some of the most unrealistic space games out there.
> people saying that getting rid of the gravity well issue for docking will kill the game should maybe look back at some of the predecessors and their design choices. because some of them thought otherwise and they werent wrong.
> I'm saying, let's make this a configurable option, and we both can have what we want. I'm even willing to bet you a whole 50 cents that in time you will see that I am right on this issue. Peace.

yes i did know all of this, and the fastest way to get out of a gravity well isnt just to enter maneuver mode but to turn on slide mode and just hit the afterburner. but the problem comes with the fact that ***none of that is fun*** and it still takes forever.
hireing someone else to bring my mule to the planet i am at can be lucrative but the point is moot when i took the mule out to go pirating and and coming in with several yeoman worth of goods i managed to loot. and then will be leaving with an empty mule to go hit more ox's or whatever.  also just paying to have the mule brought opver to you eliminates the feel of owning a mule for space trucking purposes.

as for the configurable, sure i am good for that. as for the 50 cents? nah i am having more fun on this docking refactor branch then i have had ever with vegastrike.

royfalk commented:
> Now what really does suck without any hope of getting it better in the game is the fact that planetary docks can be occupied. It would be nice if one could purchase a priority pass in the game that allows you to dock when in range, and kicks out whoever is docked at the moment. :)

This is no longer true. It was ridicules that planets had one parking spot and I replaced that with the unlimited dock size.

> In my opinion, the docking distance should be a configurable option.

It is for planets (always) and stations (simple mode).

> the thing presently killing my fun isnt even related to this it's the equipment selling bugs that are currently being experienced.

Can you please link to an issue and open one if need be with details. 

> Distance from planet at launch

This was never a feature of this PR. Whoever asked for it, please open another issue.

yowshi commented:
> > the thing presently killing my fun isnt even related to this it's the equipment selling bugs that are currently being experienced.
> 
> Can you please link to an issue and open one if need be with details.

https://github.com/vegastrike/Vega-Strike-Engine-Source/issues/1038
this thing and i may need to open another one. i havent had any time this week to do any playing so i havent checked the latest commit.

royfalk commented:
@yowshi #1038 was closed by #1047.

yowshi commented:
like i said i hadnt had a chance to try the new commit yet.

yowshi commented:
i got some time today so lets try this latest commit.
fresh built commit and fresh asset-production-master
atlantis docking range at 2500km? that's a bit shrunk but it works.
serenity seems to have a nigh infinite docking range now? or at least i can dock with serenity all the way out at atlantis. 98.27 light minutes away and i can dock with serenity.

royfalk commented:
> serenity seems to have a nigh infinite docking range now? or at least i can dock with serenity all the way out at atlantis. 

Well, that was unexpected. I'll try and test this over the weekend and see if I can figure out what changed. I've been making incrementally smaller changes. The only big one was the return to classic docking for stations.

yowshi commented:
yeah i flew out to over 5 light hours and i could still dock with a station. i dont know what the actual range is but i suspect any station would die before i found ot.

kheckwrecker commented:
> > serenity seems to have a nigh infinite docking range now? or at least i can dock with serenity all the way out at atlantis.
> 
> Well, that was unexpected. I'll try and test this over the weekend and see if I can figure out what changed. I've been making incrementally smaller changes. The only big one was the return to classic docking for stations.

I'm having this same problem. It's fun to dock with a station after launching from Atlantis. Quick way to make a lot of money when you don't actually have to fly anywhere!

kheckwrecker commented:
The other problem I'm having with this PR is that docking ranges displayed for planets are not the actual docking distance. For example, for Atlantis, the docking countdown starts at 7500 km (good) and it says Docking: Ready at 2500 km, which is good, but you can't actually dock at that distance. In fact, you can't actually dock until you reach 1000 km, while it says Docking: Ready all the way in. 

Interestingly enough, 1,000 km is the original docking distance for Atlantis, so that hasn't actually changed with this PR. 

Tested this by also flying to Phillies. I can't tell you what distance Docking: Ready first appeared, but I couldn't actually dock until 339 km out from the planet, which is again the original docking distance for Phillies.  I can test the other planets in Cephid-17, but I'm pretty confident they'll all work the same way.

yowshi commented:
> The other problem I'm having with this PR is that docking ranges displayed for planets are not the actual docking distance. For example, for Atlantis, the docking countdown starts at 7500 km (good) and it says Docking: Ready at 2500 km, which is good, but you can't actually dock at that distance. In fact, you can't actually dock until you reach 1000 km, while it says Docking: Ready all the way in.
> 
> Interestingly enough, 1,000 km is the original docking distance for Atlantis, so that hasn't actually changed with this PR.
> 
> Tested this by also flying to Phillies. I can't tell you what distance Docking: Ready first appeared, but I couldn't actually dock until 339 km out from the planet, which is again the original docking distance for Phillies. I can test the other planets in Cephid-17, but I'm pretty confident they'll all work the same way.

i was not experiencing this issue. i was easily docking around the 2400 km range.

royfalk commented:
Sorry. I was stuck trying to debug #1048 to no success. It'll have to wait until I come back a week from now.
